### PR TITLE
Chordality features added

### PR DESCRIPTION
**Pull request by Toptachamann**
New features are added to the ChordalityInspector
* Hole detection
* Graph coloring
* Maximum independent set finding
* Maximum clique finding
All algorithms run in linear time.

I need to mention one feature. I didn't find any algorithm to detect a hole in the graph, so here is the way I used to do it:
 * find the first vertex v in the order produced by the iterator whose predecessors don't form a clique.
* find a pair of vertices a, b, which are predecessors of v and are not adjacent. These three vertices belong to some chordless cycle in the G[S] where G[S] is a subgraph of G on vertices in S = {u : index_in_order(u) <= index_in_order(v)}.
* use dfs to find any cycle in G[S], in which every vertex isn't adjacent to v, except for a and b.
* find a chordless subcycle in linear time.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java [hunk 1]
```diff
@@ -42,8 +45,14 @@
  * <p>
  * Both lexicographical BFS and maximum cardinality search run in $\mathcal{O}(|V| + |E|)$. Checking whether given order
  * is the perfect elimination order via {@link ChordalityInspector#isPerfectEliminationOrder(List)} takes
- * $\mathcal{O}(|V| + |E|)$ as well. So, overall time complexity of the method
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java]: Rewrite of the javadoc:

> Tests whether a graph is <a href="https://en.wikipedia.org/wiki/Chordal_graph">chordal</a>. A chordal graph is a simple graph in which all <a href="http://mathworld.wolfram.com/GraphCycle.html">cycles</a> of four or more vertices have a <a href="http://mathworld.wolfram.com/CycleChord.html">chord</a>. A chord is an edge that is not part of the cycle but connects two vertices of the cycle.
> A graph is chordal if and only if it has a <a href="https://en.wikipedia.org/wiki/Chordal_graph#Perfect_elimination_and_efficient_recognition">perfect elimination ordering</a>. A perfect elimination ordering in a graph is an ordering of the vertices of the graph such that, for each vertex $v$, $v$ and the neighbors of $v$ that occur after $v$ in the order form a clique. This implementation uses either {@link MaximumCardinalityIterator} or {@link LexBreadthFirstIterator} to compute a perfect elimination order. The desired method is specified during construction time.
> <p>
> Chordal graphs are a subset of the <a href="http://mathworld.wolfram.com/PerfectGraph.html">perfect graphs</a>. They may be recognized in polynomial time, and several problems that are hard on other classes of graphs such as minimum vertex coloring or determining maximum cardinality cliques and independent set can be performed in polynomial time when the input is chordal.
> <p>
> All methods in this class run in $\mathcal{O}(|V| + |E|)$ time. Determining whether a graph is chordal takes $\mathcal{O}(|V| + |E|)$ time, independent of the algorithm ({@link MaximumCardinalityIterator} or {@link LexBreadthFirstIterator}) used to compute the perfect elimination order. Similarly, for chordal graphs, this class can determine a perfect elmination order, a minimum vertex coloring, a maximum cardinality clique, or an independent set in $\mathcal{O}(|V| + |E|)$ time. Finally, if the input graph is not chordal, this class can detect a chordless cycle in $\mathcal{O}(|V| + |E|)$ time.
> 
> <p>
> All the methods in this class are invoked in a lazy fashion, meaning that computations are only started once the method gets invoked.

Code version:
```
Tests whether a graph is <a href="https://en.wikipedia.org/wiki/Chordal_graph">chordal</a>. A chordal graph is a simple graph in which all <a href="http://mathworld.wolfram.com/GraphCycle.html">cycles</a> of four or more vertices have a <a href="http://mathworld.wolfram.com/CycleChord.html">chord</a>. A chord is an edge that is not part of the cycle but connects two vertices of the cycle.
A graph is chordal if and only if it has a <a href="https://en.wikipedia.org/wiki/Chordal_graph#Perfect_elimination_and_efficient_recognition">perfect elimination order</a>. A perfect elimination order in a graph is an ordering of the vertices of the graph such that, for each vertex $v$, $v$ and the neighbors of $v$ that occur after $v$ in the order form a clique. This implementation uses either {@link MaximumCardinalityIterator} or {@link LexBreadthFirstIterator} to compute a perfect elimination order. The desired method is specified during construction time.
<p>
Chordal graphs are a subset of the <a href="http://mathworld.wolfram.com/PerfectGraph.html">perfect graphs</a>. They may be recognized in polynomial time, and several problems that are hard on other classes of graphs such as minimum vertex coloring or determining maximum cardinality cliques and independent set can be performed in polynomial time when the input is chordal.
<p>
All methods in this class run in $\mathcal{O}(|V| + |E|)$ time. Determining whether a graph is chordal takes $\mathcal{O}(|V| + |E|)$ time, independent of the algorithm ({@link MaximumCardinalityIterator} or {@link LexBreadthFirstIterator}) used to compute the perfect elimination order. Similarly, for chordal graphs, this class can determine a perfect elmination order, a minimum vertex coloring, a maximum cardinality clique, or an independent set in $\mathcal{O}(|V| + |E|)$ time. Finally, if the input graph is not chordal, this class can detect a chordless cycle in $\mathcal{O}(|V| + |E|)$ time.

<p>
All the methods in this class are invoked in a lazy fashion, meaning that computations are only started once the method gets invoked.
```

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java [hunk 2]
```diff
@@ -167,45 +239,260 @@ public boolean isPerfectEliminationOrder(List<V> vertexOrder) {
...
+                isChordal();
+            }
+            if (chordal) {
+                Map<V, Integer> vertexColoring = new HashMap<>(order.size());
+                Map<V, Integer> vertexInOrder = getVertexInOrder(order);
+                for (V vertex : order) {
+                    vertexColoring.put(vertex, getPredecessors(vertexInOrder, vertex).stream().map(vertexColoring::get)
+                            .max(Integer::compareTo).orElse(-1) + 1);
+                }
+                int maxColor = order.stream().map(vertexColoring::get).max(Integer::compareTo).orElse(-1) + 1;
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java]: consider using vertexColoring.values.stream.distinct.count

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java [hunk 3]
```diff
@@ -167,45 +239,260 @@ public boolean isPerfectEliminationOrder(List<V> vertexOrder) {
...
+     * @return a maximum independent set of the {@code graph} if it is chordal, false otherwise.
+     */
+    private Set<V> lazyComputeMaximumIndependentSet() {
+        if (maximumIndependentSet == null) {
+            if (!isComputed()) {
+                isChordal();
+            }
+            if (chordal) {
+                Set<V> restricted = new HashSet<>();
+                Set<V> independent = new HashSet<>();
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java]: can be replaced by:
maximumIndependentSet=new HashSet<>();

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java [hunk 4]
```diff
@@ -167,45 +239,260 @@ public boolean isPerfectEliminationOrder(List<V> vertexOrder) {
...
+                        independent.add(previous);
+                        for (E edge : graph.edgesOf(previous)) {
+                            V opposite = Graphs.getOppositeVertex(graph, edge, previous);
+                            if (!previous.equals(opposite)) {
+                                restricted.add(opposite);
+                            }
+                        }
+                    }
+                }
+                return maximumIndependentSet = independent;
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java]: This method can be reduced in size by removing the last 2 else clauses, and retaining the `return maximumIndependentSet` at the end.

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java [hunk 5]
```diff
@@ -128,6 +167,50 @@ public boolean isChordal() {
         return lazyComputeOrder();
     }
 
+    /**
+     * Returns some chordless cycle on 4 or more vertices of the {@code graph}.
```
[the remark this answers is not in the document]
Comment (Toptachamann) [reply jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java]: I've found another definition: http://mathworld.wolfram.com/ChordlessCycle.html, which, to my mind, won't interfere with the definition of a hole we use in WeakChordalityInspector. This problem is really controversial. In [this article](http://repository.cmu.edu/cgi/viewcontent.cgi?article=1081&context=tepper) the following definition is used:
"A hole in G is an induced subgraph of G that is a cycle of length at least four...". I think the best way to deal with this issue is to use everywhere in ChordalityInspector the term "chordless cycle" and in the WeakChordalityInspector the term "hole".

[the remark this answers is not in the document]
Comment (jkinable) [reply jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChordalityInspector.java]: A hole is simply a chordless cycle. The smallest chordless cycle (hole) that exists is defined on 4 vertices. So in this PR (chordal graphs) that's fine; just use the javadoc snippet I wrote here.

In the weak chordality PR, we have to state explicitely: a hole of size 5 or larger.

**Comment by jkinable** [comment]
This looks very nice. Great job adding these features.
Make sure the class headers have the right format, in particular:
* (C) Copyright <current year>-<current year>, by <your name> and Contributors.
For format details, see [point 1](https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution#a-writing-a-new-class-or-package)

@d-michail any more feedback? At a later point in time I'd like to move the `getColoring`, `getMaximumIndependentSet` and `getMaximumIndependentSet` methods to another package. E.g. have a class `org.jgrapht.alg.color.ChordalGraphColoring` instead of a ChordalityInspector.getColoring. What do you think? Similar for the clique/independent set. Nevertheless, we would first have to create some packages and interface classes for that (we dont have a package/interface for independent sets). I'm happy to push those in a separate PR. We might be able to salvage some of the interfaces from #249

**Comment by d-michail** [comment]
Looks fine. Yes, let us do interface refactoring in another PR.

**Comment by jkinable** [comment]
@Toptachamann Thanks again for this great contribution! I've updated our [wiki](https://github.com/jgrapht/jgrapht/wiki/Open-tasks,-projects-and--collaboration-ideas#chordal-graphs) page to reflect the new changes.

**Pull request merged**